test/unit/visual/cases/shape_modes.js

```diff
@@ -0,0 +1,118 @@
+  p5 ............... The p5 Instance
+  shape ............ The shape to draw. Either 'ellipse', 'arc', or 'rect'
+  mode ............. The ellipseMode (for ellipse and arc), or rectMode (for rect)
+                     Either p5.CORNERS, p5.CORNER, p5.CENTER or p5.RADIUS
+  x1, x2, x2, y2 ... Coordinates specifying the shape in CORNERS mode,
+                     i.e. (x1, y1) and (x2, y2) specify two opposite corners P1 and P2
+*/
+function shapeCorners(p5, shape, mode, x1, y1, x2, y2) {
+  // Adjust coordinates for testing modes other than CORNERS
+  if (mode === p5.CORNER) {
+    // Find top left corner
+    let x = p5.min(x1, x2);  // x
+    let y = p5.min(y1, y2);  // y
+    // Calculate width and height
+    // Don't use abs(), so we get negative values as well
+    let w = x2 - x1; // w
+    let h = y2 - y1; // h
+    x1 = x; y1 = y; x2 = w; y2 = h;
+  } else if (mode === p5.CENTER) {
+    // Find center
+    let x = (x2 + x1) / 2; // x
+    let y = (y2 + y1) / 2; // y
+    // Calculate width and height
+    // Don't use abs(), so we get negative values as well
+    let w = x2 - x1;
+    let h = y2 - y1;
+    x1 = x; y1 = y; x2 = w; y2 = h;
+  } else if (mode === p5.RADIUS) {
+    // Find Center
+    let x = (x2 + x1) / 2; // x
+    let y = (y2 + y1) / 2; // y
+    // Calculate radii
+    // Don't use abs(), so we get negative values as well
+    let r1 = (x2 - x1) / 2; // r1;
+    let r2 = (y2 - y1) / 2; // r2
+    x1 = x; y1 = y; x2 = r1; y2 = r2;
+  }
+
+  if (shape === 'ellipse') {
+    p5.ellipseMode(mode);
+    p5.ellipse(x1, y1, x2, y2);
+  } else if (shape === 'arc') {
+    // Draw four arcs with gaps inbetween
+    const GAP = p5.radians(30);
+    p5.ellipseMode(mode);
+    p5.arc(x1, y1, x2, y2, 0 + GAP, p5.HALF_PI - GAP);
+    p5.arc(x1, y1, x2, y2, p5.HALF_PI + GAP, p5.PI - GAP);
+    p5.arc(x1, y1, x2, y2, p5.PI + GAP, p5.PI + p5.HALF_PI - GAP);
+    p5.arc(x1, y1, x2, y2, p5.PI + p5.HALF_PI + GAP, p5.TWO_PI - GAP);
+  } else if (shape === 'rect') {
+    p5.rectMode(mode);
+    p5.rect(x1, y1, x2, y2);
+  }
+}
+
+
+/*
+  Comprehensive test for rendering ellipse(), arc(), and rect()
+  with the different ellipseMode() / rectMode() values: CORNERS, CORNER, CENTER, RADIUS.
+  Each of the 3 shapes is tested with each of the 4 possible modes, resulting in 12 test.
+  Each test renders the shape in 16 different coordinate configurations,
+  testing combinations of positive and negative coordinate values.
+*/
+visualSuite('Shape Modes', function(...args) {
+  // Shapes to test
+  const SHAPES = [ 'ellipse', 'arc', 'rect' ];
+
+  // Modes to test (used with ellipseMode or rectMode, according to shape)
+  const MODES  = [ 'CORNERS', 'CORNER', 'CENTER', 'RADIUS' ];
+
+  for (let shape of SHAPES) {
+    visualSuite(`Shape ${shape}`, function() {
+
+      for (let mode of MODES) {
+        visualTest(`Mode ${mode}`, function(p5, screenshot) {
+          p5.createCanvas(240, 500);
```

Comment:
Alright, I believe creating a larger canvas might be an issue. When the tests run, a bigger canvas could make them take a bit longer due to its size. Plus, including files that are 15-20kB could make the codebase pretty heavy. Could we perhaps reduce it to 50x50? That would probably take up minimal space in bytes.

By doing this, we could include the tests one by one. For reference, you might want to check out this link: https://github.com/processing/p5.js/pull/6783/files#diff-aefa4b8e5287fbd0cdf37a6ec8d27322a4e5c893e99ec74d32053e60e2bdfd7d

Maybe by using a 50x50 size, we could adjust the parameter values of _shapeCorners accordingly?

Reply:
I think in this case using many smaller canvases instead of a few bigger ones, will actually make things slower. Because then we'll have the overhead of createCanvas() i.e. creating a rendering context much more often, and the overall codebase size will also be larger with many small PNGs because of the PNG headers.

And to thoroughly test everything, we have 4 modes, times 3 shapes, times 4 quadrants, times 4 combinations of points, equals 192 shapes to draw. I see no way around that. Bear in mind this adds 192 test for fundamental drawing operations that were previously untested.

Therefore I propose making the canvas in the tests as small as possible, while still testing all the cases. 
I have reduced the canvas size from 240x500 to 60x125 px, which reduces the total size of the generated PNGs from 223 to 12 kB.

Reply:
sounds good!